query/src/main/scala/filodb/query/exec/PeriodicSamplesMapper.scala

```diff
@@ -182,27 +182,36 @@ extends RangeVectorCursor with StrictLogging {
     // the compiler. Copy to a local variable to reduce some overhead.
     val wit = windowIt
 
-    wit.nextWindow()
-    while (wit.hasNext) {
-      val nextInfo = wit.next
-      try {
-        rangeFunction.addChunks(nextInfo.getTsVectorAccessor, nextInfo.getTsVectorAddr, nextInfo.getTsReader,
-                                nextInfo.getValueVectorAccessor, nextInfo.getValueVectorAddr, nextInfo.getValueReader,
-                                wit.curWindowStart, wit.curWindowEnd, nextInfo, querySession.queryConfig)
-      } catch {
-        case e: Exception =>
-          val tsReader = LongBinaryVector(nextInfo.getTsVectorAccessor, nextInfo.getTsVectorAddr)
-          val valReader = rv.partition.schema.data.reader(rv.valueColID,
-                                                          nextInfo.getValueVectorAccessor,
-                                                          nextInfo.getValueVectorAddr)
-          qLogger.error(s"addChunks Exception: info.numRows=${nextInfo.numRows} " +
-             s"info.endTime=${nextInfo.endTime} curWindowEnd=${wit.curWindowEnd} tsReader=$tsReader " +
-             s"timestampVectorLength=${tsReader.length(nextInfo.getTsVectorAccessor, nextInfo.getTsVectorAddr)} " +
-             s"valueVectorLength=${valReader.length(nextInfo.getValueVectorAccessor, nextInfo.getValueVectorAddr)}", e)
-          throw e
+    try {
+      wit.nextWindow()
+
+      while (wit.hasNext) {
+        val nextInfo = wit.next
+        try {
+          rangeFunction.addChunks(nextInfo.getTsVectorAccessor, nextInfo.getTsVectorAddr, nextInfo.getTsReader,
+                                  nextInfo.getValueVectorAccessor, nextInfo.getValueVectorAddr, nextInfo.getValueReader,
+                                  wit.curWindowStart, wit.curWindowEnd, nextInfo, querySession.queryConfig)
+        } catch {
+          case e: Exception =>
+            val tsReader = LongBinaryVector(nextInfo.getTsVectorAccessor, nextInfo.getTsVectorAddr)
+            val valReader = rv.partition.schema.data.reader(rv.valueColID,
+                                                            nextInfo.getValueVectorAccessor,
+                                                            nextInfo.getValueVectorAddr)
+            qLogger.error(s"addChunks Exception: info.numRows=${nextInfo.numRows} " +
+              s"info.endTime=${nextInfo.endTime} curWindowEnd=${wit.curWindowEnd} tsReader=$tsReader " +
+              s"timestampVectorLength=${tsReader.length(nextInfo.getTsVectorAccessor, nextInfo.getTsVectorAddr)} " +
+              s"valueVectorLength=${valReader.length(nextInfo.getValueVectorAccessor, nextInfo.getValueVectorAddr)}", e)
+            throw e
+        }
+      }
+      rangeFunction.apply(wit.curWindowStart, wit.curWindowEnd, sampleToEmit)
+    } catch {
+      case e: Throwable => {
+        wit.unlock()
+        wit.close()
+        throw e
       }
     }
```

Comment:
This is more of a bandaid IMO. The real fix should be to catch the exception upstream. Doesnt the "finally" catch the exception and close the cursor ?